### PR TITLE
feat: add retry and http client plugins

### PR DESCRIPTION
**Pull request by AndrewFossAWS**
## Description of changes
* Add retry and http client plugins

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt [hunk 1]
```diff
@@ -81,19 +82,13 @@ open class HttpProtocolServiceClient(
                 writer.openBlock("return ClientBuilder<\$L>(defaultPlugins: [", "])", serviceSymbol.name) {
 
                     val defaultPlugins: MutableList<Plugin> = mutableListOf(DefaultClientPlugin())
-                    val customPlugins: MutableList<Plugin> = mutableListOf()
```
[the remark this answers is not in the document]
Comment (AndrewFossAWS) [reply smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt:84]: There weren't any plugins that fell under custom plugins. It should only be default plugins.

File: Sources/ClientRuntime/Plugins/DefaultClientPlugin.swift [hunk 2]
```diff
@@ -9,10 +9,17 @@ public class DefaultClientPlugin: Plugin {
...
     public func configureClient(clientConfiguration: ClientConfiguration) {
         if var config = clientConfiguration as? DefaultClientConfiguration {
-            // Populate default values for configuration here if they are missing
             config.retryStrategyOptions =
                 DefaultSDKRuntimeConfiguration<DefaultRetryStrategy, DefaultRetryErrorInfoProvider>
                     .defaultRetryStrategyOptions
         }
+
+        if var config = clientConfiguration as? DefaultHttpClientConfiguration {
+            var httpClientConfiguration = DefaultSDKRuntimeConfiguration<DefaultRetryStrategy, DefaultRetryErrorInfoProvider>
```
[the remark this answers is not in the document]
Comment (AndrewFossAWS) [reply Sources/ClientRuntime/Plugins/DefaultClientPlugin.swift:18]: > Will the DefaultSDKRuntimeConfiguration struct used here be refactored down the line?

Yes. I will try to send out a PR later this week.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt [hunk 3]
```diff
@@ -66,7 +68,7 @@ open class HttpProtocolServiceClient(
 
     open fun renderConvenienceInitFunctions(serviceSymbol: Symbol) {
         writer.openBlock("public convenience required init() throws {", "}") {
-            writer.write("let config = try \$L(\"\$L\", \"\$L\")", serviceConfig.typeName, serviceName, serviceSymbol.name)
+            writer.write("let config = try \$L()", serviceConfig.typeName)
```
Comment (syall) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt:71]: Nice change 👍 . Are these set anywhere else right now?

**Review by syall — APPROVED** [review]

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt (hunk 3 above)
Comment (AndrewFossAWS) [reply smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolServiceClient.kt:71]: Nope :)

**Review by sichanyoo — CHANGES_REQUESTED** [review]
Just need swiftlint warnings (10) taken care of, should be good afterwards.

**Review by sichanyoo — APPROVED** [review]

**Pull request merged**